miner/collator.go

```diff
@@ -0,0 +1,312 @@
+	Service() interface{}
+}
+
+// Pool is an interface to the transaction pool
+type Pool interface {
+	Pending(bool) (map[common.Address]types.Transactions, error)
+	Locals() []common.Address
+}
+
+type BlockState interface {
+	AddTransaction(tx *types.Transaction) (error, *types.Receipt)
+	RevertTransactions(count uint) error
+	Commit() bool
+	Copy() BlockState
+	State() vm.StateReader
+	Signer() types.Signer
+	Header() *types.Header
+	Etherbase() common.Address
+}
+
+const (
+	InterruptNone int = iota
+	InterruptResubmit
+	InterruptNewHead
+)
+
+type InterruptContext interface {
+	InterruptState() int
+}
+
+type Collator interface {
+	CollateBlock(bs BlockState, ctx InterruptContext)
+	Start(pool Pool)
+	Close()
+}
+
+var (
+	ErrInterrupt          = errors.New("interrupt triggered")
+	ErrGasLimitReached    = errors.New("gas limit reached")
+	ErrNonceTooLow        = errors.New("tx nonce too low")
+	ErrNonceTooHigh       = errors.New("tx nonce too high")
+	ErrTxTypeNotSupported = errors.New("tx type not supported")
+	ErrStrange            = errors.New("strange error")
+	ErrGasFeeCapTooLow    = errors.New("gas fee cap too low")
+)
+
+func LoadCollator(filepath string, configPath string) (Collator, CollatorAPI, error) {
+	p, err := plugin.Open(filepath)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	v, err := p.Lookup("PluginConstructor")
+	if err != nil {
+		return nil, nil, errors.New("Symbol 'APIExport' not found")
+	}
+
+	pluginConstructor, ok := v.(func(config *map[string]interface{}) (Collator, CollatorAPI, error))
+	if !ok {
+		return nil, nil, errors.New("Expected symbol 'API' to be of type 'CollatorAPI")
+	}
+
+	f, err := os.Open(configPath)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer f.Close()
+
+	config := make(map[string]interface{})
+	if err := toml.NewDecoder(f).Decode(&config); err != nil {
+		return nil, nil, err
+	}
+
+	collator, collatorAPI, err := pluginConstructor(&config)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return collator, collatorAPI, nil
+}
+
+const (
+	interruptNotHandled int32 = 0
+	interruptIsHandled  int32 = 1
+)
+
+type blockState struct {
+	worker     *worker
+	env        *environment
+	start      time.Time
+	logs       []*types.Log
+	shouldSeal bool
+	snapshots  []int
+	committed  bool
+
+	// shared values between multiple copies of a blockState
+
+	interrupt *int32
+	// mutex to make sure only one blockState is calling commit at a given time
+	commitMu *sync.Mutex
+	// this makes sure multiple copies of a blockState can only trigger
+	// adjustment of the recommit interval once
+	interruptHandled *int32
+	// prevents calls to worker.commit (with a given blockState) after
+	// CollateBlock call on that blockState returns. examined in commit
+	// when commitMu is held.  modified right after CollateBlock returns
+	done *bool
+	// calling Commit() copies the value of env to this value
+	// and forwards it to the sealer via worker.commit() if shouldSeal is true
+	resultEnv *environment
```

Comment:
Is this resultEnv necessary? 
Whenever we create a blockState, we pass the generated environment to the collator and do whatever change it wants. Then we always send the deep copied environment to the commit function seems it will do some mutations inside. Finally we can either keep the env or drop it, but we don't need another field `resultEnv`.

Reply:
Perhaps in the collator, they will copy the block state and do some executions on the top and then pick a suitable one as the return? If so it makes sense to have the `resultEnv` field.

Reply:
My thinking with `resultEnv` is that as a collator `Commit()`s various blocks, the `blockState`s track the last block committed to, and after `CollateBlock`, sets the pending block as `resultEnv`.

The pending block could be set directly as part of `blockState.Commit()` but it could cause the pending block to change erratically during `CollateBlock`.  That being said, I recall hearing that the pending block isn't really used that much so perhaps it's not worth caring about this behavior.